### PR TITLE
Initial CLI command tests

### LINKED ISSUE #284: Initial CLI tests
Initial tests to
- [x] run a service using a specified image and port options
- [x] see the service listed
- [x] verify service is accessible at published ports
- [x] delete service and see the service has been removed from service list

### PR DESCRIPTION
**Pull request by JosephGJ**
Closes #284 

To test:

```
$ go test github.com/appcelerator/amp/cmd/amp/cli
```

### CODE REVIEW

**Comment by subfuzion** [comment]
@JosephGJ I rebased for you and force pushed. To update your branch:

```
$ git fetch origin cli-command-tests
$ git reset --hard origin/cli-command-tests
```

**Comment by subfuzion** [comment]
@JosephGJ I updated the test command for you.

File: cmd/amp/cli/cli_test.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+	"gopkg.in/yaml.v2"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+	"testing"
+	"time"
+
+	// Amplifier
```
Comment (subfuzion) [inline cmd/amp/cli/cli_test.go]: Comment not necessary.

File: cmd/amp/cli/cli_test.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+	"github.com/appcelerator/amp/api/rpc/stack"
+	"github.com/appcelerator/amp/api/server"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+type TestSpec struct {
```
Comment (subfuzion) [inline cmd/amp/cli/cli_test.go:14]: Don't export test types in the cli package. Better yet, rename this package to `cli_test`.

File: cmd/amp/cli/cli_test.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+	etcdDefaultEndpoints    = "http://localhost:2379"
+	serverAddress           = "localhost" + defaultPort
+	elasticsearchDefaultURL = "http://localhost:9200"
+	kafkaDefaultURL         = "localhost:9092"
+	influxDefaultURL        = "http://localhost:8086"
+	dockerDefaultURL        = "unix:///var/run/docker.sock"
+	dockerDefaultVersion    = "1.24"
+)
+
+// Amplifier
```
Comment (subfuzion) [inline cmd/amp/cli/cli_test.go]: Amplifier set up code not necessary. There is a helper function to start the server [here](https://github.com/appcelerator/amp/blob/master/api/server/server_utils.go#L76). See `TestMain` code [here](https://github.com/appcelerator/amp/blob/master/api/rpc/service/service_test.go#L92) for an example. You've already imported the package, so you should be able to just use the helper functions like the example.

File: cmd/amp/cli/cli_test.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+	if err != nil {
+		fmt.Println("connection failure")
+		os.Exit(1)
+	}
+	client = stack.NewStackServiceClient(conn)
+	os.Exit(m.Run())
+}
+
+func TestCmds(t *testing.T) {
+	parseEnv()
```
Comment (subfuzion) [inline cmd/amp/cli/cli_test.go]: You don't need this (you needed it for `TestMain`).

File: cmd/amp/cli/cli_test.go [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+	}
+	for _, spec := range commandMap {
+		cmdSplit := strings.Fields(spec.Cmd)
+		optionsSplit := []string{}
+		for _, val := range spec.Options {
+			optionsSplit = append(optionsSplit, strings.Fields(val)...)
+		}
+		commandFinal := append(cmdSplit, spec.Args...)
+		commandFinal = append(commandFinal, optionsSplit...)
+		fmt.Println(commandFinal, "Command passed")
```
Comment (subfuzion) [inline cmd/amp/cli/cli_test.go]: Don't use `fmt` for tests; instead, use `t.Log`, etc.

File: cmd/amp/cli/cli_test.go [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+		RunCmd(t, commandFinal)
+	}
+}
+
+func RunCmd(t *testing.T, cmdString []string) {
+	fmt.Println(cmdString, "Running...")
```
Comment (subfuzion) [inline cmd/amp/cli/cli_test.go]: Same: don't use `fmt`

File: cmd/amp/cli/cli_test.go [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+	cmd := exec.Command(cmdString[0], cmdString[1:]...)
+	result, err := cmd.CombinedOutput()
+	if err != nil {
+		t.Log(cmdString)
+		t.Error(err)
+		t.Fail()
+	} else {
+		fmt.Print(cmdString, " Command result:\n", string(result))
+	}
+	time.Sleep(2 * time.Second)
```
Comment (subfuzion) [inline cmd/amp/cli/cli_test.go]: Why the Sleep call? `cmd.CombinedOutput()` blocks until finished, no?

**Review by subfuzion — COMMENTED** [review]
Tests pass. Nice work @JosephGJ. I've got trivial comments for you to address, but overall, great job.

File: cmd/amp/cli/cli_test.go (hunk 7 above)
Comment (JosephGJ) [reply cmd/amp/cli/cli_test.go]: Sleep call removed. Replaced with a for loop for recursing a command that fails. Primarly for the purpose of ensure that the pinger isn't curled before pinger has appeared in the swarm.

File: cmd/amp/cli/cli_test.go (hunk 5 above)
Comment (JosephGJ) [reply cmd/amp/cli/cli_test.go]: t.Log used in every instance of a fmt.Print statement.

File: cmd/amp/cli/cli_test.go (hunk 4 above)
Comment (JosephGJ) [reply cmd/amp/cli/cli_test.go]: Removed from TestCmds

File: cmd/amp/cli/cli_test.go (hunk 3 above)
Comment (JosephGJ) [reply cmd/amp/cli/cli_test.go]: Removed unnecessary amplifier setup code. Used the StartTestServer function suggested.

File: cmd/amp/cli/cli_test.go (hunk 2 above)
Comment (JosephGJ) [reply cmd/amp/cli/cli_test.go:14]: Test types removed.

File: cmd/amp/cli/cli_test.go (hunk 1 above)
Comment (JosephGJ) [reply cmd/amp/cli/cli_test.go]: Comments removed.

**Comment by JosephGJ** [comment]
Modified to take into account tony's comments:
- Removed amplifier setup code
- fmt.Println replaced with t.Log
- Sleep call replaced with a for loop designed to recursive over a failing command.
- Unnecessary comments and types removed

Same test as before:
`$ go test github.com/appcelerator/amp/cmd/amp/cli`

Will continue to working on Cli command tests.

File: cmd/amp/cli/cli_test.go [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+		name := f.Name()
+		t.Log("Loading file:", name)
+		contents, err := ioutil.ReadFile(path.Join(testDir, name))
+		if err != nil {
+			t.Errorf("unable to load test sample: %s. Error: %v", name, err)
+		}
+		testSpec := &TestSpec{
+			fileName: name,
+			contents: contents,
+			valid:    valid,
```
Comment (subfuzion) [inline cmd/amp/cli/cli_test.go]:
```
$ go test github.com/appcelerator/amp/cmd/amp/cli
# github.com/appcelerator/amp/cmd/amp/cli_test
cmd/amp/cli/cli_test.go:59: undefined: valid
FAIL    github.com/appcelerator/amp/cmd/amp/cli [build failed]
```

**Comment by subfuzion** [comment]
Don't merge `master` (https://github.com/appcelerator/amp/pull/305/commits/64321d0e23f060639b7b62524f20196db607f014), please rebase your PRs (and squash your commits) instead.

**Review by subfuzion — APPROVED** [review]
Nice work, @JosephGJ 👍

**Pull request merged**